File: src/minecraft/net/machinemuse/api/electricity/ElectricAdapter.java
package net.machinemuse.api.electricity;

import ic2.api.IElectricItem;
import net.minecraft.item.ItemStack;
import universalelectricity.core.item.IItemElectric;

public abstract class ElectricAdapter {
	public static ElectricAdapter wrap(ItemStack stack) {
		if (stack == null) {
			return null;
		}
		if (stack.getItem() instanceof IItemElectric) {
			return new UEElectricAdapter(stack);
		} else if (stack.getItem() instanceof IElectricItem) {
			return new IC2ElectricAdapter(stack);
			// } else if (stack.getItem() instanceof IChargeableItem) {
			// return new TEElectricAdapter(stack);
			// } else if (stack.getItem() instanceof IXynergyItem) {
			// return new XyElectricAdapter(stack);
		} else {
			return null;
		}
	}

	/**
	 * Call to get the energy of an item
	 * 
	 * @param stack
	 *            ItemStack to set
	 * @return Current energy level
	 */
	public abstract double getCurrentEnergy();

	/**
	 * Call to set the energy of an item
	 * 
	 * @param stack
	 *            ItemStack to set
	 * @return Maximum energy level
	 */
	public abstract double getMaxEnergy();

	/**
	 * Call to drain energy from an item
	 * 
	 * @param stack
	 *            ItemStack being requested for energy
	 * @param requested
	 *            Amount of energy to drain
	 * @return Amount of energy successfully drained
	 */
	public abstract double drainEnergy(double requested);

	/**
	 * Call to give energy to an item
	 * 
	 * @param stack
	 *            ItemStack being requested for energy
	 * @param provided
	 *            Amount of energy to drain
	 * @return Amount of energy used
	 */
	public abstract double giveEnergy(double provided);
}

File: src/minecraft/net/machinemuse/api/electricity/ElectricAdapter.scala
package net.machinemuse.api.electricity

import net.minecraft.item.ItemStack
import universalelectricity.core.item.IItemElectric
import net.machinemuse.powersuits.common.ModCompatability
import ic2.api.item.{ElectricItem, IElectricItem}
import net.machinemuse.api.electricity.ElectricConversions._
import cofh.api.energy.IEnergyContainerItem

object ElectricAdapter {
  def wrap(stack: ItemStack): ElectricAdapter = {
    if (stack == null) return null
    stack.getItem match {
      case i: IEnergyContainerItem => new TEElectricAdapter(stack)
      case i: IItemElectric => new UEElectricAdapter(stack)
      case i: IElectricItem => if (ModCompatability.isIndustrialCraftLoaded) new IC2ElectricAdapter(stack) else null
      case _ => null
    }
  }

}

abstract class ElectricAdapter {
  def getCurrentEnergy: Double

  def getMaxEnergy: Double

  def drainEnergy(requested: Double): Double

  def giveEnergy(provided: Double): Double
}

class IC2ElectricAdapter(val stack: ItemStack) extends ElectricAdapter {
  val item = stack.getItem.asInstanceOf[IElectricItem]

  def getCurrentEnergy: Double = museEnergyFromEU(ElectricItem.discharge(stack, Integer.MAX_VALUE, getTier, true, true))

  def getMaxEnergy: Double = museEnergyFromEU(item.getMaxCharge(stack))

  def drainEnergy(requested: Double) = museEnergyFromEU(ElectricItem.discharge(stack, museEnergyToEU(requested).toInt, getTier, true, false))

  def giveEnergy(provided: Double): Double = museEnergyFromEU(ElectricItem.charge(stack, museEnergyToEU(provided).toInt, getTier, true, false))

  def getTier = item.getTier(stack)
}

class TEElectricAdapter(val stack: ItemStack) extends ElectricAdapter {
  val item = stack.getItem.asInstanceOf[IEnergyContainerItem]

  def getCurrentEnergy: Double = museEnergyFromRF(item.getEnergyStored(stack))

  def getMaxEnergy: Double = museEnergyFromRF(item.getMaxEnergyStored(stack))

  def drainEnergy(requested: Double): Double = museEnergyFromRF(item.extractEnergy(stack, museEnergyToRF(requested), true))

  def giveEnergy(provided: Double): Double = museEnergyFromRF(item.receiveEnergy(stack, museEnergyToRF(provided), true))

}


class UEElectricAdapter(val stack: ItemStack) extends ElectricAdapter {
  val item = stack.getItem.asInstanceOf[IItemElectric]

  def getCurrentEnergy: Double = museEnergyFromJoules(item.getElectricityStored(stack))

  def getMaxEnergy: Double = museEnergyFromJoules(item.getMaxElectricityStored(stack))

  def drainEnergy(requested: Double): Double = {
    val joules: Double = museEnergyToJoules(requested)
    val received: Float = item.discharge(stack, joules.toFloat, true)
    museEnergyFromJoules(received)
  }

  def giveEnergy(provided: Double): Double = {
    val joulesToProvide: Double = museEnergyToJoules(provided)
    val eatenJoules: Float = item.recharge(stack, joulesToProvide.toFloat, true)
    museEnergyFromJoules(eatenJoules)
  }
}